### PR TITLE
Fix the dead is_fast_build check so limits_fast.yaml is actually installed on Fast test

### PR DESCRIPTION
**Pull request by groeneai**
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
...

### Description

`tests/config/users.d/limits_fast.yaml` has never been installed on any CI job.
`is_fast_build` (`tests/config/install.sh:74-77`) only `return`s an exit code and prints nothing,
while the call site read stdout: `if [[ $(is_fast_build) == 1 ]]`. The substitution is always
empty, so the branch was never taken. Any job's `set -x` trace shows it (`++ return 1`,
`+ [[ '' == 1 ]]`, `grep -c limits_fast job.log` = 0), so `max_execution_time: 60` has never been
in effect, and the `CancellationChecker` it arms (`ProcessList.cpp:381`) has never run here.

The fix makes both halves agree on one convention: `is_fast_build` stays a predicate, the call
site becomes a plain `if`.

This re-lands #97823, reverted by #98026 because the LLVM coverage build passes the predicate
(`-DNDEBUG`, no `-fsanitize=`) while being ~2-3x slower. That objection was correct, so the
coverage exclusion is included, keyed on the `WITH_COVERAGE` row rather than on `CXX_FLAGS`, which
cannot see coverage flags (applied per directory, while `CXX_FLAGS` renders
`@FULL_CXX_FLAGS_NORMALIZED@`). `tests/clickhouse-test` reads the same row for the same reason.
An unanswerable probe declines; the guards compare strings, so an empty result is not a shell error.

**Impact is limited to `Fast test`,** because the gate now also requires `--fast-test`, whose only
caller is `ci/jobs/fast_test.py:330`. Over 7 days of CIDB, 0 of 53 243 160 passing test rows on
that job exceeded 60 s (slowest 57 840 ms), and structurally so: the runner already gives each
test file a 60 s budget (`--timeout 60`, the only one in `ci/`). I did not widen it, because other
jobs satisfying `is_fast_build` do run single queries past 60 s.

`ci/tests/test_install_sh_fast_limits.py` (new) drives the real `install.sh` and asserts on the
resulting symlink; nothing previously did, which is how the dead branch survived.
`ci/defs/job_configs.py` adds both files to the `CI Tests` digest, so that test is scheduled
when the gate it guards changes.

### Version info
- Merged into: `26.8.1.510` (included in `26.8` and later)

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary>Internal second-model review (4 rounds, 22 findings, 18 AGREE / 1 DISAGREE / 3 recorded-only)</summary>

Each round is an independent cold review of the resulting code plus a separate second-model pass;
findings are adjudicated against measurements, not prose.

**r1 (1 blocker, 2 major, 2 nits)**

- ❌ blocker: the blast-radius measurement rested on a 6 h window from 2026-02 in which four of
  today's heaviest tests did not exist, and which contained no `Fast test` run at all. Scope was
  narrowed to `--fast-test` and re-measured on 7 days of current data. No test was tagged,
  weakened or pinned to fit the cap.
- ⚠️ the affected-job enumeration missed `Bugfix validation (functional tests, *)`; the set is now
  derived from `FAST_TEST`, which has one setter and one caller, so it is closed by construction.
- ⚠️ nothing asserted on the symlink, which is how the dead branch survived the original revert.
  `ci/tests/test_install_sh_fast_limits.py` was added.
- 💡 two nits: an unsourced timing figure was removed, and `-eq` on a possibly-empty substitution
  became a string comparison.

**r2 (2 major, 5 nits)**

- ⚠️ the first version of the new test matched on query *text* and re-implemented the predicate in
  Python, so four production mutations left it green. It now rewrites `system.build_options` into a
  synthetic table and hands the unmodified query text to a real server, so the SQL under test is
  genuinely evaluated.
- 💡 test docstrings restated the PR narrative; trimmed.

**r3 (2 major, 2 nits)**

- ⚠️ the coverage guard failed *open*: an unanswerable `WITH_COVERAGE` probe fell through to the
  `CXX_FLAGS` check, which by construction cannot identify a coverage build, so it would report
  "fast" and re-create the regression that caused the original revert. It now requires an explicit
  non-coverage answer and declines otherwise, which degrades to the pre-change behaviour.
- ⚠️ the new test could not protect the file it guards: at that point `CI Tests` digested `./ci`
  only, and that same path list drives both changed-file filtering and result caching, so an
  `install.sh`-only change would have skipped the job. `install.sh` and `limits_fast.yaml` were
  added to that digest.
- 💡 two nits: the call-site comment overstated when the runner kills a timed-out test file, and a
  test count was stale.

**r4 (2 major, 4 nits); the second-model pass returned zero findings**

- ⚠️ two description-drift findings, both against this description rather than the code: it did not
  mention `ci/defs/job_configs.py` at all, though that file is a third of the diff, and it described
  the coverage guard only in the exclude direction, omitting the fail-closed behaviour that two of
  the eleven tests pin. Both are fixed above.
- 💡 the only DISAGREE of the review: the three storage exclusions inside `is_fast_build`
  (S3 / azure / encrypted) cannot fire at the sole call site, since the only `--fast-test` caller
  exports none of those variables. They are kept deliberately: they are the exclusions requested in
  the #98026 revert, they defend the predicate's contract if a second caller appears, and one is
  pinned by a test. Recorded rather than removed, and left out of the description to keep it on the
  merge decision.
- 💡 recorded agreements, no change: the invariant-carrier enumeration was re-derived independently
  and matched (single consumer; both CI coverage builds covered; no other `return $(...)` construct
  under `tests/` or `ci/`; both sibling predicates already use the correct convention), and the 60 s
  bound was re-measured independently (0 of 53 243 160 passing `Fast test` rows over 7 days above
  60 s, slowest 57 840 ms). Worth adding that `max_execution_time: 60` was unconditionally active
  for every job until Apr 2025, when `9d2563d9a6c2` moved it into `limits_fast.yaml`, so this
  re-lands a strictly narrower version of a previously global setting.

Validation carried forward: 11 tests, all passing; 12 production mutations, each caught by a named
test.

</details>

**Comment by groeneai** [comment]
<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|----------|--------|
| a | **Deterministic repro?** | Yes. `set -x` trace of `install.sh` with a `PATH` wrapper forcing the `system.build_options` answer: on `origin/master` with a fast-build answer the symlink is not created (`++ is_fast_build` / `++ return 1` / `+ [[ '' == 1 ]]`), matching the trace in any real job log. 100% deterministic - a shell branch either runs or it does not. The same comparison runs in CI as `ci/tests/test_install_sh_fast_limits.py`. |
| b | **Root cause explained?** | Yes. Shell `return` sets `$?` and writes nothing to stdout, but the call site consumed stdout via `$(is_fast_build)`. The substitution is always empty, so `[[ '' == 1 ]]` is always false and `ln -sf .../limits_fast.yaml` never ran on any job. The polarity is also inverted for a naive `if`: the SQL predicate returns 1 for "is a fast build" while `return 1` means failure. |
| c | **Fix matches root cause?** | Yes. The two halves are made to agree on one calling convention: `is_fast_build` stays a predicate (0 = true) and the call site becomes an `if`. Nothing is widened, disabled or tagged, and no test is edited to accommodate the newly-live profile. The coverage exclusion that the #98026 revert asked for is added inside the predicate, where a caller cannot bypass it. The `FAST_TEST` condition answers a separate question the original one-line gate never asked - which jobs should get this profile - and is backed by measurement rather than assumption. |
| d | **Test intent preserved / new tests added?** | Yes. No existing test file is modified or weakened; the change restores an intended safety net. `ci/tests/test_install_sh_fast_limits.py` is new: 11 tests driving the real `install.sh` (the shell function under test is not stubbed) and asserting on the resulting symlink and its target, run by the existing `CI Tests` job. Nothing previously asserted on this symlink, which is how the dead branch survived from #98026 until now. The stub supplies the `system.build_options` rows and forwards the query text to a real `clickhouse local`, so the SQL predicates are genuinely evaluated rather than modelled. Every guard is proven load-bearing by 12 mutations, all of which fail the suite: reverting the call site, dropping the `FAST_TEST` condition, removing any one of the three guards, restoring `-eq`, five SQL mutations (dropping `upper`, reducing the coverage check to row presence, dropping the `-DNDEBUG` conjunct, flipping the sanitizer polarity, and weakening the conjunction to a disjunction), and reverting the coverage guard's fail direction. The `CI Tests` digest also lists `tests/config/install.sh`, so a later change to the gate schedules the job that runs this test instead of resolving a cached success. |
| e | **Both directions demonstrated?** | Yes. `origin/master` script + forced fast build + `--fast-test` = no symlink; this PR + the same forced answer = `+ ln -sf .../limits_fast.yaml ...` and the symlink present. Traces captured for both, plus a direction proving the scope narrowing (without `--fast-test` the `&&` short-circuits and `is_fast_build` never runs). |
| f | **Fix is general across code paths?** | Yes. `git grep -n is_fast_build` returns exactly the definition (`install.sh:74`) and the single call site, so the convention change cannot affect any other consumer. The coverage check is in the predicate rather than the caller precisely so it also holds for `tests/docker_scripts/stress_tests.lib:72`, which invokes `install.sh` without `--llvm-coverage`. All three callers of `tests/config/install.sh` were enumerated with the reason each is or is not affected (`ci/jobs/fast_test.py:330`, `ci/jobs/functional_tests.py:771`, `stress_tests.lib:72`); no integration job installs these configs at all. The sibling predicates in the same file were read too: `is_sanitizer_build` (`:225`) already uses the correct string-comparison shape, and `is_openssl_fips_build` (`:253`) still uses `-eq` - the same nit, but on a pre-existing line this PR does not touch, so it is left for its own change rather than bundled here. |
| g | **Fix generalizes across inputs (params/datatypes/wrappers)?** | Yes. Exercised across build kind as real `CXX_FLAGS` strings evaluated by a real server rather than as precomputed verdicts (fast `-O3 -DNDEBUG`, sanitizer `-O2 -DNDEBUG -fsanitize=address,undefined`, debug `-O0 -g` with no `-DNDEBUG`) and coverage as a real `WITH_COVERAGE` row value, storage environment (`S3`, Azure, encrypted, none), the exported-versus-set distinction for `USE_ENCRYPTED_STORAGE`, and the presence or absence of `--fast-test`. Value tolerance matches the existing in-tree readers: `upper(value) NOT IN ('ON', '1')` excludes `ON`/`on`/`1`, the same values `tests/clickhouse-test:4923-4929` and `tests/integration/helpers/cluster.py:5032-5035` already accept, and the lowercase case is pinned by a test. Boundary cases measured and pinned: an absent `WITH_COVERAGE` row and an absent `CXX_FLAGS` row both return an empty string, both are handled as strings rather than integers, and both decline - the coverage probe is written so an unanswerable result declines, because `CXX_FLAGS` cannot identify a coverage build and so cannot compensate for it. |
| h | **Backward compatible? (maintainer-approved exception only)** | Yes, trivially. No setting default changes, no `SettingsChangesHistory.cpp` entry, no serialization or format change, nothing user-facing. The diff only changes which test config file is symlinked in CI, and only on the Fast test job. |
| i | **Invariants and contracts preserved?** | Yes. Invariant: `limits_fast.yaml` is installed exactly on jobs where a 60 s per-query cap cannot fire on a healthy test. That is measured, not assumed: on `Fast test` 0 of 53 243 160 passing rows over 7 days exceed 60 s (slowest 57 840 ms), and structurally cannot, because the runner's own per-file `--timeout 60` binds tighter than a per-query limit; on the other jobs `is_fast_build` accepts it is false, which is why the scope is narrowed. `is_fast_build` honours the shell predicate contract on every path - the three guards `return 1` explicitly and the final string comparison is the function's exit status. `set -e` safety holds because the function and the `&&` are used only as an `if` condition, where POSIX suspends `set -e`, and the early returns are inside the function so they cannot abort the script. Both probes share one fail direction: an unanswerable probe declines, so the predicate degrades to the previous no-symlink behaviour rather than capping a build it could not classify. `bash -n` clean, including under all twelve mutations. |

</details>

Session id: cron:clickhouse-review-slot-48:20260729-025400

File: ci/tests/test_install_sh_fast_limits.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+"""
```
Comment (Algunenano) [inline ci/tests/test_install_sh_fast_limits.py]: Please remove this test, it's useless

**Comment by groeneai** [comment]
### CI finish ledger - f51e578db4ec670894b08f76c241e3f887bc0b75

Every failure below has an owner: a fixing PR (ours or external), or a full-effort fix task
whose fixing-PR link will be posted here when it opens. Only `CH Inc sync` is exempt.

| Check / test | Reason | Owner / fixing PR |
|---|---|---|
| Stateless tests (amd_llvm_coverage, ParallelReplicas, s3 storage, parallel) / 01666_merge_tree_max_query_limit | the per-table concurrency throttle the test asserts is keyed on query id, and secondary parallel-replicas reads are sent without one, so the holder query competes against itself (3/3 reruns, deterministic) | #112385 (ours, open) |

Session id: cron:our-pr-ci-monitor:20260729-120000

File: ci/tests/test_install_sh_fast_limits.py (hunk 1 above)
Comment (Algunenano) [reply ci/tests/test_install_sh_fast_limits.py]: @groeneai pls

Comment (groeneai) [reply ci/tests/test_install_sh_fast_limits.py]: Removed in bb14a17f93f494497fea77dde8bca897c7f44ba6. The `ci_tests` digest entries for `install.sh` and `limits_fast.yaml` only existed to schedule that test, so they went with it and the PR is now just the `install.sh` fix. That leaves the gate with no automated coverage; let me know if you want any replacement.

File: tests/config/install.sh [hunk 2]
```diff
@@ -270,7 +286,10 @@ ln -sf $SRC_PATH/users.d/limits.yaml $DEST_SERVER_PATH/users.d/
 if check_clickhouse_version 26.1; then
     ln -sf $SRC_PATH/users.d/distributed_index_analysis.yaml $DEST_SERVER_PATH/users.d/
 fi
-if [[ $(is_fast_build) == 1 ]]; then
+# Only the Fast test job, whose runner already gives each test file a 60 second wall-clock
+# budget, so a 60 second per-query limit cannot be the first thing to fire on a healthy
+# test there. Other jobs that satisfy is_fast_build run the long tests that Fast test skips.
+if [ "$FAST_TEST" == "1" ] && is_fast_build; then
```
[the remark this answers is not in the document]
Comment (groeneai) [reply tests/config/install.sh:292]: The 126 second figure is the outer watchdog, not the per-test deadline.

`clickhouse-test` bounds each test file twice. The inner bound is `proc.wait(args.timeout)` at `tests/clickhouse-test:3516` in `run_single_test`; `process_result_impl` then marks the file FAIL / `Timeout!` because that wait left `proc.returncode is None` (`:3149`, `:3158`, `:1297`). With `--timeout 60` that is 60 seconds per test file, and `args.timeout` is never reassigned in the file. The `signal.alarm(int(args.timeout * 1.1) + 60)` you quote (`:4630-4631`, `run_tests_array`) wraps `test_case.run` plus `process_result`, so it covers that inner wait plus stdout post-processing, `_cleanup` (up to another 60 seconds, `:3822`) and the stacktrace dumps. It fires only if the Python side hangs.

A query therefore cannot outlive the file issuing it: the query reaches 60 seconds at `file_start + d + 60` with `d >= 0`, while the wait expires at `file_start + 60`. The runner fires first, so the profile can only change the reported reason of an already-failing test, never turn a passing one red.

Runtimes agree. Over 30 days of `Fast test%`, 214M test runs: p99 4.3 s, p999 9.7 s, 488 runs at or above 60 s, 437 of those in the 60-126 s band, and exactly 1 of the 437 is OK (`01824_prefer_global_in_and_join`, `Fast test (arm_darwin)`, 60060 ms, a boundary race). Bucketed over 50-200 s: 50 s 179/172 OK, 60 s 436/1, 80 s 1/0, 120 s 7/0, 130 s 2/0. Durations pile up against the 60 s wall, not 126 s. Fast test also passes `--no-long`.

Keeping the cap. The comment's "60 second wall-clock budget" is that `proc.wait`, so it is accurate as written; happy to reword it to name `proc.wait` explicitly if that reads better.

**Review by Algunenano — APPROVED** [review]
